RandLAPACK/drivers/rl_bqrrp.hh

```diff
@@ -18,6 +18,11 @@ namespace RandLAPACK {
 template <typename T, typename RNG>
 class BQRRPalg {
     public:
+        struct SubroutineType{
+            enum qrcp_wide_subroutines {luqr, geqp3};
+            enum qr_tall_subroutines {geqrt, cholqr, geqrf};
+            enum apply_trans_q_subroutines {ormqr, gemqrt};
+        };
```

Comment:
I suggest
```c++
struct Options {
    struct QRCPWide    { enum Enum { LUQR,  QP3           }; };
    struct QRTall      { enum Enum { QRT,   CholQR, QRF   }; };
    struct ApplyTransQ { enum Enum { ORMQR, GEMQRT        }; };
};
```
You could change ``Options`` to ``Subroutines`` if you'd like.

Reply:
Any particular reason for that change or is it just your stylistic preference?

Reply:
1. We use CamelCase for type names.
2. It's shorter.

Reply:
AIt, will rework everything then

Reply:
Downside: three "using" statements instead of one in each file.

Reply:
If you're using them multiple times in one file I don't think that's a significant downside. You can also just do
```c++
using BQO = BQRRP::Options;

// ... later ...
auto luqr = BQO::QRCPWide::LUQR; 
auto qrf = BQO::QRTall::QRF;
```

Reply:
I feel like using three things instead of one twice (three structs instead of one and three "using" statements instead of one) is still suboptimal, but if you say it is best, I'll do it.

Reply:
The advantage of the nested version is that function signatures accepting these arguments become clearer.

For example,
```c++
template <typename T>
void bqrrp_functional(
    int64_t m, int64_t n, T* A, int64_t lda, T* tau, int64_t* piv,
    Options::QRTall::Enum      qr_tall       = Options::QRTall::QRF,
    Options::QRCPWide::Enum    qrcp_wide     = Options::QRCPWide::LUQR, 
    Options::ApplyTransQ::Enum apply_trans_q = Options::ApplyTransQ::ORMQR
) {
    // Step 1. Construct the BQRRP object and an RNGState
    RNGState s(0);
    // Step 2. invoke the BQRRP object's "call" function.
    // Step 3. ... return.
}
```
That said, there are alternatives here. Maybe ...
```c++
struct BQRRPOptions {
    struct QRCPWide {enum Enum {LUQR, QP3}; };
    struct QRTall   {enum Enum {CholQR, QRF}; };
    QRCPWide::Enum qrcp_wide = QRCPWide::LUQR;
    QRTall::Enum qr_tall = QRTall::QRF;
};
// ^ Can define that inside or outside of the class. Note that it's a type which contains structs and can be instantiated.

template <typename T>
void bqrrp_functional(
    int64_t m, int64_t n, T* A, int64_t lda, T* tau, int64_t* piv, BQRRPOptions ops = {}
) {
    // Step 1. Construct the BQRRP object and an RNGState
    RNGState s(0);
    // Step 2. invoke the BQRRP object's "call" function.
    // Step 3. ... return.
}

// assume you have some A, piv, m, n, and tau defined ...
bqrrp_functional(m, n, A, m, tau, piv);
// ^ the compiler injects an argument BQRRPOptions ops = {}, which
//   which has ops.qrcp_wide and ops.qr_tall automatically initialized to their defaults.

using Options = BQRRPOptions;
// ^ Can include that statement in places where it's the meaning of "Options" is unambiguous

bqrrp_functional(m, n, A, m, tau, {Options::QRCPWide::QP3, Options::QRTall::CholQR}); 
// ^ being explicit about options without constructing a BQRRPOptions object on another line.

```

Reply:
@TeachRaccooon you can mess with alternatives here: 

https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXAEx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApJIBCZ86SX1kBPAMqN0AYVS0AriwZ6XAGTwGTAA5bwAjTGIJAE5SAAdUBUJHBncvHz1E5IcBQOCwlkjorjjbTHtUoQImYgJ0718uG0w7XIZq2oJ80Iio2JsauobM5oUh7qDeov7SgEobVE9iZHYOEw0AQQ3N8eJPewBqcwBFACUz5QB5ePaFQ5MAdkstw7fDvYOCQ/PXZQB1PBYB7PRjeQ7YBjgp7mfxyc6kH7KADMTwAIiZkeYQRisTt3h8CPsjucACpiWjvGFglgQqG0mGuBAeBE/M4AMXRmOxXLxr3evwBQPYIEh4IAjit4gB9ADuwoeyLRbL%2BgKwIBAcPO3PxArO5NotA1YtpkulNUNiuVZIpGvOnL520euJe2y2nmSRkONzuVuO50uPtSCh1WyDAnuqHiIee4YYCjtZ1Vwo1WrOiLjCZANsNGqZLLOvNdOyC33Q3hYAE8IKcLtdbsHDlH7pjlTD0XMQa6CcRMARlgxDhpQ06MVsS4JDgA3MTSrh%2B8ssKsQOYj0vT2eSBcV6swzOJ5Pq7MojMNiOJg1GkD52ja52rx0cBZGjgAVl4vg4WlIqE4risVgfEsKyYA8kjIjwpAEJoT4LAgmBMFg0QrqQADWIDIjEAB0XCvhoABs%2BEABwxDEyL4RoGgxAALPonDUbwLASJRpCft%2Bv4cLwCYaFBMELHAsAwIgKCoCw8R0FE5CUGgYkSdEwBcIpfB0AQUQJhA4QwaQ4RBLUlacJBOnMMQlZXOE2gVNB3C8DJbCCFcDC0PpX68Fg4SeMArgUgm1mkFgLCGMA4guX5eC9pUU6YD536YKoFSeKpBm8KWrRabQeDhMQenuOqvlEngTG%2BZFxDhEkmBopgAVGOlRh8XwBjAAoABqeCYLKNyMElMiCCIYjsFwPH8IIigqOoIW6M0Bi1aYFhWPoGUJpACxRncnC8KgxXEMKi0oeUlROBALgjE0pABFMhTFFkSQpAIx1XTkqQ9BdswtG0VQTHdYytJZ71dE9fQlIMXSfUDdT/TMJQLAowGrBIz6cO%2BrFaRxhyqER%2BEALT4dRhzAMgyCHIpOGHBAuCECQYEQXMvBWVo/FIDJ4n0GQFAQIzckgApSk0LQqnEOpmkhUZeldcLJlmRZ9hdbZjAEA5TlaW5HleYaPmQf5gXBd%2B%2BDhQ4kXRbwsXxYlvkpS%2B37pZl2UYGskH5YVkHFaVSgVVVQVBKALkLFQDXNa17XxJ1vlDcIojiAN3XyEoahabo0hTcYAEWPN4Q7ctZ7xmtP6bdt8BQ99%2B2%2BIdDBuB4jR%2BCX4OXc02Q3WkZejAk13tFXL17e0nTDA3J3t79YPnQDejjMD3dDxMreQ4syyw1w8Nvh%2ByOcKj6NYzjeME0T86k/gRDEJTs803xCzoXhdEcAxpBMafbHrVnCbNLTT6kAJwlLAQ8QJVJh07yQQJ6CHvVxBEUjiNGO409CyiyvEOqL5EY3x/JwK4CUP7fFQFQZemNsa43xoTLgxMIDuFkszfe1NeJezgghJClA54XyvjxeBHEuIgB4o/OYx8QCSEeFhV8uFSKPHwo8ZEBFXyCORGfZEC8QqMLIXTZ%2BQkIAMxtmTXeX8eaMFYLbXgACw79VoiHUBY1vxx30IFFASdrC9wOkdUezQzoFEHqUJuD1bo2KcXXCeQ8C4dw%2Bq4yxHRx4DwhgMYeXcMg9wCfYoJs8p4gU4AAejiYcUklZA6HErMBQ4aBgTwV7IiCmtBUCIUOIYQ4sVWDxHoFhCc3wFDik8LUTAEB1z0k7O2fkbxez9mIIOekhwABUhx6Qji5OOLYnwjhtM2AScZ3xBRqlAjCUpvSTRdjTIie0RZNntMJMSWZ%2BoKRdhpHSaEzwbysg2c6HUlzHQEjmSmUUvTJTIBlPKLAI5bn7NzA8iUxBzQUmGc6b0GcFCbF5lwYZYyiRfC7LqN4Ry7nAhhGstknJrndnePCz5lJGTMlvOmFFWypl6kPKBJ5LzhTvL1JeQ4ZoLS0ABcqTMoKCCSBHE/WBkj2J/nMUBaeCzwKSEPuQtCGFkRYWRBKyVUqpVnwvgwu%2BzCZGwTPoKpGUis6sIWMVZIThqJAA%3D

Reply:
And it becomes clearer because the user knows that we're using an enumerations?
Thats it?

Reply:
What would your version of the following signature look like if we used your preferred method?
```c++
template <typename T>
void bqrrp_functional(
    int64_t m, int64_t n, T* A, int64_t lda, T* tau, int64_t* piv,
    Options::QRTall::Enum      qr_tall       = Options::QRTall::QRF,
    Options::QRCPWide::Enum    qrcp_wide     = Options::QRCPWide::LUQR, 
    Options::ApplyTransQ::Enum apply_trans_q = Options::ApplyTransQ::ORMQR
);
```

Reply:
Also, idk if there's any sense in making a separate function that allows for tuning optional stuff, since it is optional.
The user may want to choose one of the subroutine types to be changed from the default, while not willing to think at all about the other subroutine types.

Reply:
The plan for RandLAPACK has always been to have an object-oriented API and a procedural API. I just want to know what the procedural API would look like with your approach.

Reply:
Wouldnt it just be

```
template <typename T>
void bqrrp_procedural_api(
    int64_t m, int64_t n, T* A, int64_t lda, T* tau, int64_t* piv,
    Options::QRTall      qr_tall       = Options::QRTall::QRF,
    Options::QRCPWide    qrcp_wide     = Options::QRCPWide::LUQR, 
    Options::ApplyTransQ apply_trans_q = Options::ApplyTransQ::ORMQR
);
```

if i were to change names?

Reply:
Please verify on Godbolt.

Reply:
Idk how that platform works, but i just added this into one of RandLAPACK tests and it compiles just fine.
The syntax is the same as what I've pointed out, but the names are obviously different.

```
...
    template <typename T, typename RNG>
    static void buffer1(
        RandLAPACK::BQRRP<T, RNG>::SubroutineType::qr_tall_subroutines qr_tall) {
            int a = 0;
        }

    template <typename T, typename RNG>
    static void buffer2(
        RandLAPACK::BQRRP<T, RNG>::SubroutineType::qr_tall_subroutines qr_tall = RandLAPACK::BQRRP<T, RNG>::SubroutineType::qr_tall_subroutines::cholqr) {
            int a = 0;
        }
};

TEST_F(TestBQRRP, test_buf) {
    RandLAPACK::BQRRP<double, r123::Philox4x32>::SubroutineType::qr_tall_subroutines qr_tall = RandLAPACK::BQRRP<double, r123::Philox4x32>::SubroutineType::qr_tall_subroutines::cholqr;
    buffer1<double, r123::Philox4x32>(qr_tall);
    buffer2<double, r123::Philox4x32>();
}
```

Reply:
Looks like the compiler is indeed happy with this
```c++
struct Options {
    enum QRCPWide {LUQR, QP3};
    enum QRTall {QRT, CholQR, QRF};
    enum ApplyTransQ {ORMQR, GEMQRT};
};

// below is just a demo.
template <typename T>
int bqrrp_procedural_api(
    int m, int n, T* A, int lda, T* tau, int* piv,
    Options::QRTall      qr_tall       = Options::QRTall::QRF,
    Options::QRCPWide    qrcp_wide     = Options::QRCPWide::LUQR, 
    Options::ApplyTransQ apply_trans_q = Options::ApplyTransQ::ORMQR
) {
    return 0;
}
```
Suggestion: define the struct type outside the BQRRPalg class and name it BQRRPOptions. Then at the top of the BQRRPalg class have a line
```c++
using Options = BQRRPOptions;
```
If you take this approach then the BQRRP implementation can use the shorter Options symbol, while test code can refer to BQRRPOptions like ``RandLAPACK::BQRRPOptions::QRTall::CholQR`` instead of ``RandLAPACK::BQRRP<double, r123::Philox4x32>::Options::QRTall::CholQR``.

Reply:
Yep, which is the same as what I've shown, just different names. 

Can do, I just thought that we wanted this struct to be formally attached to the BQRRP class.
Then again, since the struct was public, that doesn't make too much difference. 
Will apply this change a bit later.

Reply:
@rileyjmurray wait a sec, your initial complain about the way the enums were set was related to the fact that my enums pollute the RandLAPACK namespace and now you're suggesting we throw them away from the BQRRP class, which, technically, also pollutes the RandLAPACK namespace.

Reply:
You're right to ask this question.

In an ideal world the enum(s) would live inside BQRRP. The problem is that since BQRRP is templated we can't just write ``BQRRP::Options::QRTall::CholQR``. Even if we used a default template parameter for the RNG we'd have to specify the template parameter for the numerical precision. That's awkward because the enum has no semantic tie to the numerical precision.

Having a single BQRRPOptions struct -- which contains all enums inside of it -- lets us group these options together without needing to specify unrelated template parameters of a BQRRP class.

Reply:
Got it, just pushed the change to that effect; the names are different because "Subroutines" instead of "Options" matches the paper contents, and the function names are lowercase full function names because that is cleaner for the user and I use lowercase in other enums in RandLAPACK.